src/backend/distributed/master/master_repair_shards.c

```diff
@@ -238,45 +313,282 @@ RecreateTableDDLCommandList(Oid relationId)
+	SearchShardPlacementInList(shardPlacementList, sourceNodeName, sourceNodePort);
+}
+
+
+/*
+ * CopyShardPlacement copies a shard along with its co-located shards from a source node
+ * to target node. CopyShardPlacement does not make any checks about state of the shards.
+ * It is caller's responsibility to make those checks if they are necessary. If
+ * dropTargetPlacement is set to true, this function will remove existing placement row
+ * then re-creates it with FILE_FINALIZED state.
  */
-static bool
-CopyDataFromFinalizedPlacement(Oid distributedTableId, int64 shardId,
-							   ShardPlacement *healthyPlacement,
-							   ShardPlacement *placementToRepair)
+static void
+CopyShardPlacement(int64 shardId, char *sourceNodeName, int32 sourceNodePort,
+				   char *targetNodeName, int32 targetNodePort, bool dropTargetPlacement)
 {
-	const char *shardTableName = NULL;
-	const char *shardQualifiedName = NULL;
-	StringInfo copyRelationQuery = makeStringInfo();
-	List *queryResultList = NIL;
-	bool copySuccessful = false;
-
-	char *relationName = get_rel_name(distributedTableId);
-	Oid shardSchemaOid = get_rel_namespace(distributedTableId);
-	const char *shardSchemaName = get_namespace_name(shardSchemaOid);
-
-	AppendShardIdToName(&relationName, shardId);
-	shardTableName = quote_identifier(relationName);
-	shardQualifiedName = quote_qualified_identifier(shardSchemaName, shardTableName);
-
-	appendStringInfo(copyRelationQuery, WORKER_APPEND_TABLE_TO_SHARD,
-					 quote_literal_cstr(shardQualifiedName), /* table to append */
-					 quote_literal_cstr(shardQualifiedName), /* remote table name */
-					 quote_literal_cstr(healthyPlacement->nodeName), /* remote host */
-					 healthyPlacement->nodePort); /* remote port */
-
-	queryResultList = ExecuteRemoteQuery(placementToRepair->nodeName,
-										 placementToRepair->nodePort,
-										 NULL, /* current user, just data manipulation */
-										 copyRelationQuery);
-	if (queryResultList != NIL)
+	ShardInterval *shardInterval = LoadShardInterval(shardId);
+	List *colocatedShardList = ColocatedShardIntervalList(shardInterval);
+	ListCell *colocatedShardCell = NULL;
+	List *ddlCommandList = NIL;
+
+	foreach(colocatedShardCell, colocatedShardList)
+	{
+		ShardInterval *colocatedShard = (ShardInterval *) lfirst(colocatedShardCell);
+		List *colocatedShardDdlList = NIL;
+
+		colocatedShardDdlList = CopyShardCommandList(colocatedShard, sourceNodeName,
+													 sourceNodePort);
+
+		ddlCommandList = list_concat(ddlCommandList, colocatedShardDdlList);
+	}
+
+	HOLD_INTERRUPTS();
+
+	SendCommandListInSingleTransaction(targetNodeName, targetNodePort, ddlCommandList);
+
+	foreach(colocatedShardCell, colocatedShardList)
+	{
+		ShardInterval *colocatedShard = (ShardInterval *) lfirst(colocatedShardCell);
+		uint64 colocatedShardId = colocatedShard->shardId;
+
+		/*
+		 * If we call this function for repair purposes, we should remove old shard
+		 * placement row.
+		 */
+		if (dropTargetPlacement)
+		{
```

Comment:
This is not dropping actual placement. Here we remove old placement row for target node and recreate it with state FILE_FINALIZED. Therefore this belongs to here. Actual dropping logic is already in another function DropShardPlacement, which is used in master_move.